internal/common_context.go

```diff
@@ -0,0 +1,216 @@
+//	  option.Query("name", "Name", param.Default("hey"))
+//	)
+func (c CommonContext[B]) QueryParam(name string) string {
+	_, ok := c.OpenAPIParams[name]
+	if !ok {
+		slog.Warn("query parameter not expected in OpenAPI spec", "param", name, "expected_one_of", c.OpenAPIParams)
+	}
+
+	if !c.UrlValues.Has(name) {
+		defaultValue, _ := c.OpenAPIParams[name].Default.(string)
+		return defaultValue
+	}
+	return c.UrlValues.Get(name)
+}
+
+func (c CommonContext[B]) QueryParamIntErr(name string) (int, error) {
+	param := c.QueryParam(name)
+	if param == "" {
+		defaultValue, ok := c.OpenAPIParams[name].Default.(int)
+		if ok {
+			return defaultValue, nil
+		}
+
+		return 0, QueryParamNotFoundError{ParamName: name}
+	}
+
+	i, err := strconv.Atoi(param)
+	if err != nil {
+		return 0, QueryParamInvalidTypeError{
+			ParamName:    name,
+			ParamValue:   param,
+			ExpectedType: "int",
+			Err:          err,
+		}
+	}
+
+	return i, nil
+}
+
+type QueryParamNotFoundError struct {
+	ParamName string
+}
+
+func (e QueryParamNotFoundError) Error() string {
+	return fmt.Errorf("param %s not found", e.ParamName).Error()
+}
+
+type QueryParamInvalidTypeError struct {
+	ParamName    string
+	ParamValue   string
+	ExpectedType string
+	Err          error
+}
+
+func (e QueryParamInvalidTypeError) Error() string {
+	return fmt.Errorf("param %s=%s is not of type %s: %w", e.ParamName, e.ParamValue, e.ExpectedType, e.Err).Error()
+}
+
+// QueryParamsArr returns an slice of string from the given query parameter.
+func (c CommonContext[B]) QueryParamArr(name string) []string {
+	_, ok := c.OpenAPIParams[name]
+	if !ok {
+		slog.Warn("query parameter not expected in OpenAPI spec", "param", name)
+	}
+	return c.UrlValues[name]
+}
+
+// QueryParamInt returns the query parameter with the given name as an int.
+// If it does not exist, it returns the default value declared in the OpenAPI spec.
+// For example, if the query parameter is declared as:
+//
+//	fuego.Get(s, "/test", myController,
+//	  option.QueryInt("page", "Page number", param.Default(1))
+//	)
+//
+// and the query parameter does not exist, it will return 1.
+// If the query parameter does not exist and there is no default value, or if it is not an int, it returns 0.
+func (c CommonContext[B]) QueryParamInt(name string) int {
+	param, err := c.QueryParamIntErr(name)
+	if err != nil {
+		return 0
+	}
+
+	return param
+}
+
+// QueryParamBool returns the query parameter with the given name as a bool.
+// If the query parameter does not exist or is not a bool, it returns the default value declared in the OpenAPI spec.
+// For example, if the query parameter is declared as:
+//
+//	fuego.Get(s, "/test", myController,
+//	  option.QueryBool("is_ok", "Is OK?", param.Default(true))
+//	)
+//
+// and the query parameter does not exist in the HTTP request, it will return true.
+// Accepted values are defined as [strconv.ParseBool]
+func (c CommonContext[B]) QueryParamBoolErr(name string) (bool, error) {
```

Comment:
I don't think these can be `internal` our users use these?

Reply:
They're embedded in a struct that is hidden behind the Context interface!

Reply:
Oh. I see now. Thanks for the clarification. I didn't realize those would be accessible though. TIL